src/main/java/com/capitalone/dashboard/collector/CollectorTaskWithGenericItem.java

```diff
@@ -0,0 +1,67 @@
+package com.capitalone.dashboard.collector;
+
+import com.capitalone.dashboard.model.Collector;
+import com.capitalone.dashboard.model.GenericCollectorItem;
+import com.capitalone.dashboard.model.relation.RelatedCollectorItem;
+import com.capitalone.dashboard.repository.CollectorItemRepository;
+import com.capitalone.dashboard.repository.GenericCollectorItemRepository;
+import com.capitalone.dashboard.repository.RelatedCollectorItemRepository;
+import com.google.common.collect.Lists;
+import org.bson.types.ObjectId;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.stereotype.Component;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Component
+public abstract class CollectorTaskWithGenericItem<T extends Collector> extends CollectorTask<T> {
+
+    private final CollectorItemRepository collectorItemRepository;
+    private final GenericCollectorItemRepository genericCollectorItemRepository;
+    private final RelatedCollectorItemRepository relatedCollectorItemRepository;
+
+    @Autowired
+    public CollectorTaskWithGenericItem(TaskScheduler taskScheduler, String collectorName, CollectorItemRepository collectorItemRepository, GenericCollectorItemRepository genericCollectorItemRepository, RelatedCollectorItemRepository relatedCollectorItemRepository) {
+        super(taskScheduler, collectorName);
+        this.collectorItemRepository = collectorItemRepository;
+        this.genericCollectorItemRepository = genericCollectorItemRepository;
+        this.relatedCollectorItemRepository = relatedCollectorItemRepository;
+    }
+
+
+    public Map<ObjectId, Set<ObjectId>> processGenericItems(List<String> toolServers) {
+        List<GenericCollectorItem> genericCollectorItems = genericCollectorItemRepository.findAllByToolNameAndProcessTimeEquals(getCollector().getName(), 0L);
+        Map<ObjectId, Set<ObjectId>> collectorItemBuildIds = new HashMap<>();
+        genericCollectorItems.forEach(gci -> {
+            toolServers.stream()
+                    .map(server -> Lists.newArrayList(collectorItemRepository.findAllByOptionMapAndCollectorIdsIn(getGenericCollectorItemOptions(server, gci), Lists.newArrayList(getCollector().getId()))))
```

Comment:
Can  `collectorItemRepository.findAllByOptionMapAndCollectorIdsIn` ever return null?

Reply:
+1

Reply:
it does not, but it could change in spring querydsl !

Reply:
See this test case for empty repository:
```
    @Test
    public void processGenericItemsEmptyGenericItemRepo() throws IOException {
        reset();

        Collector collector = new Collector();
        collector.setId(new ObjectId("5ba136290be2d32568777fa8"));
        collector.setName("Sonar");
        Map<ObjectId, Set<ObjectId>> objectIdSetMap = testCollectorTaskWithGenericItem.processGenericItems(Lists.newArrayList("http://localhost:9000"));
        List<RelatedCollectorItem> relatedCollectorItems = Lists.newArrayList(relatedCollectorItemRepository.findAll());
        assertEquals(0, objectIdSetMap.size());
        assertEquals(0, relatedCollectorItems.size());
    }
```